otherlibs/unix/envir.c

```diff
@@ -13,18 +13,56 @@
 /*                                                                        */
 /**************************************************************************/
 
+#include <caml/config.h>
+
+#ifdef HAS_UNISTD
+#include <unistd.h>
+#endif
+#include <sys/types.h>
+#ifdef HAS_GETAUXVAL
+#include <sys/auxv.h>
+#endif
+
 #include <caml/mlvalues.h>
 #include <caml/alloc.h>
 
-#ifndef _WIN32
 extern char ** environ;
-#endif
 
-CAMLprim value unix_environment(value unit)
+CAMLprim value unix_environment_unsafe(value unit)
 {
   if (environ != NULL) {
     return caml_copy_string_array((const char**)environ);
   } else {
     return Atom(0);
   }
 }
+
+static char **secure_environ(void)
+{
+#ifdef HAS_GETAUXVAL
+  if (getauxval(AT_SECURE) == 0)
+    return environ;
+  else
+   return NULL;
+#elif defined(HAS_ISSETUGID)
+  if (!issetugid ())
```

Comment:
looks good, but having two styles for comparing results (`f() == 0` and `!f()`) within 5 lines is slightly confusing